src/renderer/sw_engine/tvgSwShape.cpp

```diff
@@ -244,112 +236,26 @@ static void _dashMoveTo(SwDashStroke& dash, uint32_t offIdx, float offset, const
 }
 
 
-static void _trimPattern(SwDashStroke* dash, const RenderShape* rshape, float length, float trimBegin, float trimEnd)
+static SwOutline* _genDashOutline(const RenderShape* rshape, const Matrix& transform, SwMpool* mpool, unsigned tid, bool trimmed)
 {
-    auto begin = length * trimBegin;
-    auto end = length * trimEnd;
-
-    //default
-    if (end > begin) {
-        if (begin > 0.0f) dash->cnt = 4;
-        else dash->cnt = 2;
-        //looping
-    } else dash->cnt = 3;
-
-    if (dash->cnt == 2) {
-        dash->pattern[0] = end - begin;
-        dash->pattern[1] = length - (end - begin);
-    } else if (dash->cnt == 3) {
-        dash->pattern[0] = end;
-        dash->pattern[1] = (begin - end);
-        dash->pattern[2] = length - begin;
-    } else {
-        dash->pattern[0] = 0;     //zero dash to start with a space.
-        dash->pattern[1] = begin;
-        dash->pattern[2] = end - begin;
-        dash->pattern[3] = length - end;
+    PathCommand* cmds = nullptr;
+    Point* pts = nullptr;
+    uint32_t cmdCnt, ptsCnt;
+
+    if (trimmed) rshape->stroke->trim.trim(rshape->path.cmds, rshape->path.pts, cmds, cmdCnt, pts, ptsCnt);
```

Comment:
I looks the trimming is performed twice for the shape and stroke.... I think we can reduce to 1?